src/hotspot/share/oops/instanceKlass.cpp

```diff
@@ -211,31 +211,39 @@ bool InstanceKlass::has_nest_member(JavaThread* current, InstanceKlass* k) const
   return false;
 }
 
-// Called to verify that k is a permitted subclass of this class
-bool InstanceKlass::has_as_permitted_subclass(const InstanceKlass* k) const {
+// Called to verify that k is a permitted subclass of this class.
+// The incoming stringStream is used to format the messages for error logging and for the caller
+// to use for exception throwing.
+bool InstanceKlass::has_as_permitted_subclass(const InstanceKlass* k, stringStream& ss) const {
   Thread* current = Thread::current();
   assert(k != nullptr, "sanity check");
   assert(_permitted_subclasses != nullptr && _permitted_subclasses != Universe::the_empty_short_array(),
          "unexpected empty _permitted_subclasses array");
 
   if (log_is_enabled(Trace, class, sealed)) {
     ResourceMark rm(current);
-    log_trace(class, sealed)("Checking for permitted subclass of %s in %s",
+    log_trace(class, sealed)("Checking for permitted subclass %s in %s",
                              k->external_name(), this->external_name());
   }
 
   // Check that the class and its super are in the same module.
   if (k->module() != this->module()) {
-    ResourceMark rm(current);
-    log_trace(class, sealed)("Check failed for same module of permitted subclass %s and sealed class %s",
-                             k->external_name(), this->external_name());
+    ss.print("Failed same module check: subclass %s is in module \"%s\" for loader %s "
+             "and sealed class %s is in module \"%s\" for loader %s",
+             k->external_name(), k->module()->name_as_C_string(), k->module()->loader_data()->loader_name_and_id(),
+             this->external_name(), this->module()->name_as_C_string(), this->module()->loader_data()->loader_name_and_id());
```

Comment:
This looks quite good except that the word "for" in "for loader %s" looks a bit strange.  In API docs we speak of "the ClassLoader for a module", not the other way around. Class loaders optionally have a name so I suppose the "loader %s" should put quotes around the name like it does for the module name.

Reply:
I will change to "with loader %s".

Note that `loader_name_and_id()` already puts (single) quotes around the name of the loader. I will use single-quotes for the module name too.

Reply:
This looks good, I hadn't spotted that loader_name_and_id() returns the quotes.